File: Claude/Python/2024/Day12b_Solution_2024.py
from collections import deque

with open("Day12_Input.txt") as f:
    grid = [line.rstrip('\n') for line in f]

rows, cols = len(grid), len(grid[0])
visited = [[False] * cols for _ in range(rows)]

def in_region(r, c, ch):
    return 0 <= r < rows and 0 <= c < cols and grid[r][c] == ch

def count_corners(r, c, ch):
    # Number of corners == number of sides for a region
    corners = 0
    # Check each of the 4 corner positions (pairs of orthogonal dirs + diagonal)
    for dr, dc in [(-1, 0), (0, 1), (1, 0), (0, -1)]:
        # orthogonal neighbors: (dr,dc) and rotated 90 degrees (dc,-dr)
        n1 = in_region(r + dr, c + dc, ch)
        n2 = in_region(r + dc, c - dr, ch)
        diag = in_region(r + dr + dc, c + dc - dr, ch)
        # convex corner: both orthogonal neighbors outside
        if not n1 and not n2:
            corners += 1
        # concave corner: both orthogonal neighbors inside but diagonal outside
        elif n1 and n2 and not diag:
            corners += 1
    return corners

total = 0
for sr in range(rows):
    for sc in range(cols):
        if visited[sr][sc]:
            continue
        ch = grid[sr][sc]
        queue = deque([(sr, sc)])
        visited[sr][sc] = True
        area = 0
        sides = 0
        while queue:
            r, c = queue.popleft()
            area += 1
            sides += count_corners(r, c, ch)
            for dr, dc in [(-1,0),(1,0),(0,-1),(0,1)]:
                nr, nc = r+dr, c+dc
                if 0 <= nr < rows and 0 <= nc < cols and not visited[nr][nc] and grid[nr][nc] == ch:
                    visited[nr][nc] = True
                    queue.append((nr, nc))
        total += area * sides

print(total)

File: Claude/Rust/2024/Day12b_Solution_2024.rs
use std::collections::VecDeque;
use std::fs;

fn main() {
    let input = fs::read_to_string("Day12_Input.txt").expect("Failed to read input file");
    let grid: Vec<Vec<u8>> = input.lines().filter(|l| !l.is_empty()).map(|l| l.bytes().collect()).collect();
    let rows = grid.len();
    let cols = grid[0].len();
    let mut visited = vec![vec![false; cols]; rows];

    let in_region = |r: i64, c: i64, ch: u8| -> bool {
        r >= 0 && c >= 0 && r < rows as i64 && c < cols as i64
            && grid[r as usize][c as usize] == ch
    };

    let mut total = 0u64;

    for sr in 0..rows {
        for sc in 0..cols {
            if visited[sr][sc] {
                continue;
            }
            let ch = grid[sr][sc];
            let mut region = Vec::new();
            let mut queue = VecDeque::new();
            queue.push_back((sr, sc));
            visited[sr][sc] = true;

            while let Some((r, c)) = queue.pop_front() {
                region.push((r, c));
                for (dr, dc) in [(-1i64, 0), (1, 0), (0, -1i64), (0, 1)] {
                    let nr = r as i64 + dr;
                    let nc = c as i64 + dc;
                    if in_region(nr, nc, ch) && !visited[nr as usize][nc as usize] {
                        visited[nr as usize][nc as usize] = true;
                        queue.push_back((nr as usize, nc as usize));
                    }
                }
            }

            let area = region.len() as u64;
            let mut corners = 0u64;

            for (r, c) in &region {
                let r = *r as i64;
                let c = *c as i64;
                for (dr, dc) in [(-1i64, -1i64), (-1, 1), (1, -1), (1, 1)] {
                    let n1 = in_region(r + dr, c, ch);
                    let n2 = in_region(r, c + dc, ch);
                    let d  = in_region(r + dr, c + dc, ch);
                    // convex corner: both orthogonal neighbors outside
                    if !n1 && !n2 {
                        corners += 1;
                    }
                    // concave corner: both orthogonal neighbors inside, diagonal outside
                    if n1 && n2 && !d {
                        corners += 1;
                    }
                }
            }

            total += area * corners;
        }
    }

    println!("{}", total);
}
